[JIT] Improve class type annotation inference

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45940 [JIT] Improve class type annotation inference**

**Summary**
In `try_ann_to_type`, if an annotation has an attribute named
`__torch_script_class__`, it is assumed to be a TorchScript class that
has already been scripted. However, if it is a class that extends
another class, this code path causes a crash because it looks up the
JIT type for the class by name in the compilation unit. This JIT type
obviously cannot exist because inheritance is not supported.

This commit fixes this by looking up the qualified name of a class
in torch.jit._state._script_class in order to ascertain whether it has
already been scripted (instead of looking for a `__torch_script_class__`
attribute on the class object.

**Test Plan**
This commit adds a unit test consisting of the code sample from the
issue that reported this problem.

**Fixes**
This commit fixes #45860.

Differential Revision: [D24310027](https://our.internmc.facebook.com/intern/diff/D24310027)

Commit:
[JIT] Improve class type annotation inference

**Summary**
In `try_ann_to_type`, if an annotation has an attribute named
`__torch_script_class__`, it is assumed to be a TorchScript class that
has already been scripted. However, if it is a class that extends
another class, this code path causes a crash because it looks up the
JIT type for the class by name in the compilation unit. This JIT type
obviously cannot exist because inheritance is not supported.

**Test Plan**
This commit adds a unit test consisting of the code sample from the
issue that reported this problem.

**Fixes**
This commit fixes #45860.

[ghstack-poisoned]

eellison commented on torch/jit/annotations.py:
```diff
@@ -320,7 +320,7 @@ def try_ann_to_type(ann, loc):
         return EnumType(_qualified_name(ann), get_enum_value_type(ann, loc), list(ann))
     if inspect.isclass(ann):
-        if hasattr(ann, "__torch_script_class__"):
+        if hasattr(ann, "__torch_script_class__") and len(ann.mro()) <= 2:
```
Won't this break `torch.no_grad` ? I think it would be more intuitive to switch to storing map of {python class pointer, jit class} instead of annotating the class

SplitInfinity replied:
We have a map from `qualified_name` -> `cls` (Python class, not JIT type) in `_state.py` for all classes that have truly been scripted; what if we changed this to
```suggestion
        if torch.jit._state._get_script_class(ann):
```
In the example that motivated this fix, `A` would be in `_state._script_classes`, but `B` would not.

eellison replied:
yea sgtm

SplitInfinity replied:
Although I realize now that we're in a weird place where inheritance is prohibited if the class is scripted directly, but allowed if it is scripted implicitly (e.g. through `try_ann_to_type`) and doesn't use any features of inheritance (like `super()`).

Commit:
Update on "[JIT] Improve class type annotation inference"


**Summary**
In `try_ann_to_type`, if an annotation has an attribute named
`__torch_script_class__`, it is assumed to be a TorchScript class that
has already been scripted. However, if it is a class that extends
another class, this code path causes a crash because it looks up the
JIT type for the class by name in the compilation unit. This JIT type
obviously cannot exist because inheritance is not supported.

This commit fixes this by looking up the qualified name of a class
in torch.jit._state._script_class in order to ascertain whether it has
already been scripted (instead of looking for a `__torch_script_class__`
attribute on the class object.

**Test Plan**
This commit adds a unit test consisting of the code sample from the
issue that reported this problem.

**Fixes**
This commit fixes #45860.

[ghstack-poisoned]

eellison commented on torch/jit/annotations.py:
```diff
@@ -320,12 +321,13 @@ def try_ann_to_type(ann, loc):
         return EnumType(_qualified_name(ann), get_enum_value_type(ann, loc), list(ann))
     if inspect.isclass(ann):
-        if hasattr(ann, "__torch_script_class__"):
-            return ClassType(_qualified_name(ann))
```
grep `__torch_script_class__`: https://github.com/pytorch/pytorch/search?q=__torch_script_class__

There are a couple other places we have to update this

SplitInfinity replied:
Fixed

eellison reviewed (commented):
Looks good, two other invocations we need to update i think

Commit:
Update on "[JIT] Improve class type annotation inference"


**Summary**
In `try_ann_to_type`, if an annotation has an attribute named
`__torch_script_class__`, it is assumed to be a TorchScript class that
has already been scripted. However, if it is a class that extends
another class, this code path causes a crash because it looks up the
JIT type for the class by name in the compilation unit. This JIT type
obviously cannot exist because inheritance is not supported.

This commit fixes this by looking up the qualified name of a class
in torch.jit._state._script_class in order to ascertain whether it has
already been scripted (instead of looking for a `__torch_script_class__`
attribute on the class object.

**Test Plan**
This commit adds a unit test consisting of the code sample from the
issue that reported this problem.

**Fixes**
This commit fixes #45860.

[ghstack-poisoned]

SplitInfinity commented:
The test failure is a mypy type check failure in FX.

eellison reviewed (approved):
LGTM
